OSSM-8735 Document migration of different gateway deployments (#88648)
openshift/openshift-docs: modules/ossm-migrating-gateways-canary.adoc

// Module included in the following assemblies:
//
// * service-mesh-docs-main/migrating/migrating-gateways/ossm-migrating-gateways-assembly.adoc

:_mod-docs-content-type: PROCEDURE
[id="ossm-migrating-gateways-canary_{context}"]
= Gateway canary migration

The gateway canary migration scenario is recommend for a gradual rollout. It uses multiple gateway versions with maximum control over gateway rollout.

[NOTE]
====

> The label for the gateway namespace differs between mulitenant and cluster-wide meshes. Return to the either the multitenant or cluster-wide migration guide to understand which labels to use in your specific migration case, and to see if the `maistra.io/ignore-namespace: "true"` label should be removed.

Reviewer: ```suggestion
The label for the gateway namespace differs between mulitenant and cluster-wide meshes. Return to the either the multitenant or cluster-wide migration guide to understand which labels to use in your specific migration case.
```

====

.Procedure

. Label the gateway namespace to ensure injection from the new mesh is enabled for the namespace.

> . Add the `maistra.io/ignore-namespace: "true"` label.

Reviewer: It's imported to have all steps together, please follow what we have in the eng doc - https://github.com/openshift-service-mesh/sail-operator/blob/main/docs/ossm/ossm2-migration/gateway-migration.md#gateway-migration-guide

// Module included in the following assemblies:
//
// * service-mesh-docs-main/migrating/migrating-gateways/ossm-migrating-gateways-assembly.adoc

:_mod-docs-content-type: PROCEDURE
[id="ossm-migrating-gateways-canary_{context}"]
= Gateway canary migration

The gateway canary migration method is recommend for a gradual rollout. It uses multiple gateway versions with maximum control over your gateway rollout.

[NOTE]
====
The label for the gateway namespace differs between mulitenant and cluster-wide meshes. Return to the either the "Multitenant migration guide" or the "Cluster-wide migration guide" to understand which labels to use in your specific migration case, and to understand why the `maistra.io/ignore-namespace: "true"` label is needed.
====

.Procedure

> . Label the gateway namespace to ensure injection from the new mesh is enabled for the namespace, ensuring to add the `maistra.io/ignore-namespace: "true"` label as well as remove `istio-injection=enabled` if needed. For example:

Author: @FilipB

+
[source,terminal]
----
$ oc label namespace ${APP_NAMESPACE} istio.io/rev=${ISTIO_REVISION} maistra.io/ignore-namespace="true"
----

> 

Author: @FilipB

. Deploy a canary gateway:
+
.Example yaml for canary gateway
[source,yaml]
----
apiVersion: apps/v1
kind: Deployment
metadata:
  name: istio-ingressgateway-canary
  namespace: istio-ingress <1>
  spec:
     selector:
       matchLabels:
         istio: ingressgateway <2>
     template:
       metadata:
         annotations:
           inject.istio.io/templates: gateway
         labels:
           istio: ingressgateway
           istio.io/rev: canary <3>
       spec:
         containers:
         - name: istio-proxy
           image: auto
----
<1> Must be in in the same namespace as your existing gateway
<2> Must match your existing gateway service selector.
<3> Set to your {SMProduct} 3.0 control plane revision

. Ensure that the new gateway deployment is running with the new revision and is handling requests:
+

> .. Check that pods are running and ready.

Reviewer 2: ```suggestion
.. Check that pods are running and in the `Ready` status.
```

I believe there must be `oc get pods` command to go with this step. Please confirm with the SME/QE.

.. Check that the gateway is running the new revision by running the following command:
+
[source,terminal]
----
$ oc `istioctl ps -n istio-ingress`
----

> .. Test a sample route through the gateway.

Author: Per Dev, it is user and environment dependent.

> .. Test a sample route through the gateway.

Reviewer 2: Can you please check with Dev if we can have an example command or step to add for test application-specific routes?

// Module included in the following assemblies:
//
// * service-mesh-docs-main/migrating/migrating-gateways/ossm-migrating-gateways-assembly.adoc

:_mod-docs-content-type: PROCEDURE
[id="ossm-migrating-gateways-canary_{context}"]
= Gateway canary migration

The gateway canary migration method is recommend for a gradual rollout. It uses multiple gateway versions with maximum control over your gateway rollout.

[NOTE]
====

> The label for the gateway differs between multitenant and cluster-wide meshes. Return to the either the multitenant or cluster-wide migration guide to understand which labels to use in your specific migration case, and to check if you need to remove the `maistra.io/ignore-namespace: "true"` label.

Reviewer: ```suggestion
The label for the gateway differs between multitenant and cluster-wide meshes. Return to the either the multitenant or cluster-wide migration guide to understand which labels to use in your specific migration case, and to understand why the `maistra.io/ignore-namespace: "true"` label is needed.
```

// Module included in the following assemblies:
//
// * service-mesh-docs-main/migrating/migrating-gateways/ossm-migrating-gateways-assembly.adoc

:_mod-docs-content-type: PROCEDURE
[id="ossm-migrating-gateways-canary_{context}"]
= Gateway canary migration

Use the gateway canary migration method when you want a gradual rollout. It uses multiple gateway versions with maximum control over your gateway rollout.

[NOTE]
====
The label for the gateway namespace differs between mulitenant and cluster-wide meshes. To understand which labels to use in your specific migration case and the `maistra.io/ignore-namespace: "true"` label requirement, see the "Multitenant migration guide" or the "Cluster-wide migration guide" section.
====

.Procedure

> . Run the following command to label the gateway namespace to ensure that gateway injection is enabled from the new mesh and add the `maistra.io/ignore-namespace: "true"` label:

Reviewer 3: Not going to stop the show over this one point, but maybe you could break this up into two sentences somehow?
```suggestion
. Run the following command to label the gateway namespace and add the `maistra.io/ignore-namespace: "true"` label.
This ensures that gateway injection is enabled from the new mesh.
```

Author: Slacked about this.

+
[source,terminal]
----
$ oc label namespace <gateway_namespace> istio.io/rev=<istio_revision_name> maistra.io/ignore-namespace="true"
----
+

> This command also removes the `istio-injection=enabled` label, if needed.

Reviewer: This is not true. Please be careful not to change the logic of those steps.

. Deploy a canary gateway by using the following example:
+
.Example YAML for canary gateway
[source,yaml]
----
apiVersion: apps/v1
kind: Deployment
metadata:
  name: istio-ingressgateway-canary
  namespace: istio-ingress <1>
  spec:
     selector:
       matchLabels:
         istio: ingressgateway <2>
     template:
       metadata:
         annotations:
           inject.istio.io/templates: gateway
         labels:
           istio: ingressgateway
           istio.io/rev: canary <3>
       spec:
         containers:
         - name: istio-proxy
           image: auto
----
<1> Your `Deployment` resource must be in in the same namespace as your existing gateway.

> <2> Your `spec.selector.matchlabels.istio` must match your existing gateway service selector.

Reviewer 3: This should have a noun after the monospace text for clarity, I'm guessing this should be parameter:
```suggestion
<2> Your `spec.selector.matchlabels.istio` parameter must match your existing gateway service selector.
```

<3> Set your {SMProduct} 3.0 control plane revision as the value of the `istio.io/rev` label.

. Ensure that the new gateway deployment is running with the new revision and is handling requests:
+
.. Check that pods are running and in the `Ready` status.

.. Check that the gateway is running the new revision by running the following command:
+
[source,terminal]
----
$ oc istioctl ps -n istio-ingress
----

.. Test a sample route through the gateway.

. Gradually shift traffic between deployments:
+
.. Increase replicas for new gateway by running the following command:
+
[source,terminal]
----
$ oc scale -n istio-ingress deployment/<new_gateway_deployment> --replicas <new_number_of_replicas>
----

.. Decrease replicas for old gateway by running the following command:
[source,terminal]
+
----
$ oc scale -n istio-ingress deployment/<old_gateway_deployment> --replicas <new_number_of_replicas>
----

> . Repeat step 4, incrementally adjusting replica counts until the new gateway handles all traffic to the gateway Service.

Reviewer 3: While I couldn't find a style guide rule on this, I think it's generally best to avoid referencing specific steps, especially by number. If you had to keep this the same way, you could rephrase this as "Repeat the previous step", but that this encounters the same issues as referencing a number IMO (accuracy breaking if steps are added in this procedure).

Since this is referencing the previous steps, I think this could be simply be rewritten to be a part of step 4:
```suggestion
.. Repeat this process, incrementally adjusting replica counts until the new gateway handles all traffic to the gateway Service.
```
